squander/partitioning/tools.py

```diff
@@ -0,0 +1,189 @@
+
+        param_order (List[int] ) Tuples specifying new parameter positions: source_idx, dest_idx, param_count
+    
+    Return:
+
+        Returns with the reordered Reordered parameter array
+    """ 
+    reordered = np.empty_like(params)
+
+    for s_idx, n_idx, n_params in param_order:
+        reordered[n_idx:n_idx + n_params] = params[s_idx:s_idx + n_params]
+
+    return reordered
+
+
+
+def build_dependency(c: Circuit) -> Tuple[Dict[int, Gate], Dict[int, Set[int]], Dict[int, Set[int]]]:
+    """
+    Build dependency graphs for circuit gates.
+    
+    Args:
+        c: SQUANDER Circuit
+        
+    Returns:
+        tuple: (gate_dict, forward_graph, reverse_graph)
+    """
+    gate_dict = {i: gate for i, gate in enumerate(c.get_Gates())}
+    g, rg = {i: set() for i in gate_dict}, {i: set() for i in gate_dict}
+    
+    for gate in gate_dict:
+        for child in c.get_Children(gate_dict[gate]):
+            g[gate].add(child)
+            rg[child].add(gate)
+    
+    return gate_dict, g, rg
+
+
+def qiskit_to_squander_name(qiskit_name):
+    name = qiskit_name.upper()
+    if name == "CX":
+        return "CNOT"
+    elif name == "U":
+        return "U3"
+    elif name == "TDG":
+        return "Tdg"
+    else:
+        return name
+
+def gate_desc_to_gate_index(circ, preparts):
+    gate_dict, g, rg = build_dependency(circ)
+    L, S = [], {m for m in rg if len(rg[m]) == 0}
+    
+    curr_partition = set()
+    curr_idx = 0
+    total = 0
+    parts = [[]]
+
+    while S:
+        Scomp = {(frozenset(get_qubits(gate_dict[x])), gate_dict[x].get_Name()): x for x in S}
+        rev_Scomp = { y: x for x, y in Scomp.items()}
+        n = next(iter(Scomp.keys() & preparts[len(parts)-1]), None)
+        if n is not None: n = Scomp[n]
+
+        if n is None:
+            total += len(parts[-1])
+            curr_partition = set()
+            parts.append([])
+            n = next(iter(Scomp.keys() & preparts[len(parts)-1]), None)
+            if n is not None: n = Scomp[n]
+        
+        preparts[len(parts)-1].remove(rev_Scomp[n])
+        parts[-1].append(n)
+        curr_partition |= get_qubits(gate_dict[n])
+        curr_idx += gate_dict[n].get_Parameter_Num()
+
+        # Update dependencies
+        L.append(n)
+        S.remove(n)
+        assert len(rg[n]) == 0
+        for child in set(g[n]):
+            g[n].remove(child)
+            rg[child].remove(n)
+            if not rg[child]:
+                S.add(child)
+
+    # Add the last partition
+    total += len(parts[-1])
+    assert total == len(gate_dict)
+    # print(parts)
+    return parts
+
+def get_qiskit_partitions(filename, max_partition_size):
+    circ, parameters, qc = utils.qasm_to_squander_circuit(filename, True)
+    pm = PassManager([
+        CollectMultiQBlocks(max_block_size=max_partition_size),
+    ])
+
+    pm.run(qc)
+    blocks = pm.property_set['block_list'] # is not in topological order
```

Comment:
I think we found alas it is in topological order, it was merely that when using gate name, and qubit indexes to describe the structure, the multiplicity could be more than 1, so a set wont work but a multiset/bag (more elegant and efficient) or a list (which seems practically okay) will do the trick.